src/Uno.UI/UI/Xaml/DragDropExtension.macOS.cs

```diff
@@ -0,0 +1,178 @@
+#if __MACOS__
+#nullable enable
+
+using System;
+using System.Numerics;
+using System.Threading;
+using AppKit;
+using Uno.Extensions;
+using Uno.Logging;
+using Windows.UI.Core;
+using Windows.UI.Xaml;
+using Windows.UI.Xaml.Input;
+using Point = Windows.Foundation.Point;
+using UIElement = Windows.UI.Xaml.UIElement;
+
+namespace Windows.ApplicationModel.DataTransfer.DragDrop.Core
+{
+	internal class MacOSDragDropExtension : IDragDropExtension
+	{
+		private readonly long _fakePointerId = Pointer.CreateUniqueIdForUnknownPointer();
+		private readonly DragDropManager _manager;
+		private Uno.UI.Controls.Window _window;
+
+		public MacOSDragDropExtension(DragDropManager owner)
+		{
+			_manager = (DragDropManager)owner;
+			_window = (Uno.UI.Controls.Window)CoreWindow.GetForCurrentThread()!._window;
+
+			_window.DraggingEnteredAction = OnDraggingEnteredEvent;
+			_window.DraggingUpdatedAction = OnDraggingUpdated;
+			_window.DraggingExitedAction = OnDraggingExited;
+			_window.PerformDragOperationAction = OnPerformDragOperation;
+			_window.DraggingEndedAction = OnDraggingEnded;
+		}
+
+		private NSDragOperation OnDraggingEnteredEvent(NSDraggingInfo draggingInfo)
+		{
+			var source = new DragEventSource(_fakePointerId, draggingInfo, _window);
+			var data = DataPackage.CreateFromNativeDragDropData(draggingInfo);
+			var allowedOperations = DataPackageOperation.Copy; // Should match with return value
+			var info = new CoreDragInfo(source, data, allowedOperations);
+
+			CoreDragDropManager.GetForCurrentView()?.DragStarted(info);
+
+			// Note: No need to _manager.ProcessMove, the DragStarted will actually have the same effect
+
+			// To allow the macOS drag to continue, we must assume drop is supported here.
+			// This will immediately be updated within the OnDraggingUpdated method.
+			return NSDragOperation.Copy;
+		}
+
+		private NSDragOperation OnDraggingUpdated(NSDraggingInfo draggingInfo)
+		{
+			var operation = _manager.ProcessMoved(new DragEventSource(_fakePointerId, draggingInfo, _window));
+			return ToNSDragOperation(operation);
+		}
+
+		private void OnDraggingExited(NSDraggingInfo draggingInfo)
+		{
+			_ = _manager.ProcessAborted(new DragEventSource(_fakePointerId, draggingInfo, _window));
+			return;
+		}
+
+		private bool OnPerformDragOperation(NSDraggingInfo draggingInfo)
+		{
+			var operation = _manager.ProcessDropped(new DragEventSource(_fakePointerId, draggingInfo, _window));
+			return (operation != DataPackageOperation.None);
+		}
+
+		private void OnDraggingEnded(NSDraggingInfo draggingInfo)
+		{
+			return;
+		}
+
+		public void StartNativeDrag(CoreDragInfo info)
+			=> CoreDispatcher.Main.RunAsync(CoreDispatcherPriority.High, async () =>
+			{
+				try
+				{
+					NSDraggingSource source = new NSDraggingSource();
+					NSEvent sourceEvent = new NSEvent();
+
```

Comment:
The native dragging session is started from the ContentView insead of the NSWindow. [NSWindow.Drag](https://developer.apple.com/documentation/appkit/nswindow/1419224-drag) does not seem to be supported in Xamarin -- this method does not exist in Xamarin.mac. Attempts at invoking the objective-c code directly are so far unsuccessful.

Reply:
If no one knows of a solution to this the 'hack' I can think of would be to abort the drag within Uno as soon as it is handed off to the system. This means either (1) Dragging back into Uno wouldn't be possible or (2) Dragging back into Uno would create a new dragging session as if it started externally. I think if (2) is possible it is the best that can be done and much better than leaving two dragging sessions active: one inside Uno, one externally.

Reply:
Just had a look in the macOS contract, did you notice that the [`NSDraggingSource` is a protocol since macOS 10.7](https://developer.apple.com/documentation/appkit/nsdraggingsource) and it does have 3 methods `draggingSession` for "_endedAt_", "_movedTo_" and "_willBeginAt_" which looks like to match our needs.

Unfortunately the Xamarin's macOS binding does reflect the old contract (which is no longer present in apple doc but can be found in the [archived documentation](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/DragandDrop/Concepts/dragsource.html)) that supports only dragging images. 

I would assume that if you create your own source which inherits from the `NSDraggingSource` (you can also try to create a new one which inherits from `NSObject` and implements the [`INSDraggingSource`](https://docs.microsoft.com/en-us/dotnet/api/appkit.insdraggingsource?view=xamarin-mac-sdk-14) but sounds more risky), then you should be able to implements those methods by adding the `Export` attribute as you did for the Dragging* methods on the `Window`.

Otherwise I would try to go for option (2), the (1) wouldn't be really user friendly :/

Reply:
I did look at DraggingSource at one point -- clearly not enough :) I missed those methods so thanks for bringing it to my attention! It would make sense if that solves this.

Reply:
As usual what should have been a quick change turned into a few hours of trial and error. The [current macOS docs](https://developer.apple.com/documentation/appkit/nsdraggingsource) appear to be wrong or are somehow incompatible with Xamarin. The following code never works. I tried almost everything.

```csharp
private class BindableNSDraggingSource : NSDraggingSource
{
	[Export("draggingSession:willBeginAt:")]
	public void DraggingSessionWillBegin(NSDraggingSession session, CGPoint screenPoint)
	{
		return;
	}

	[Export("draggingSession:endedAt:operation:")]
	public void DraggingSessionEnded(NSDraggingSession session, CGPoint screenPoint, NSDragOperation operation)
	{
		return;
	}
}
 ```

However, what does work are the old methods -- still present for backwards compatibility. I found the original signatures [here](http://mirror.informatimago.com/next/developer.apple.com/documentation/Cocoa/Reference/ApplicationKit/ObjC_classic/Protocols/NSDraggingSource.html) which corresponds to the [legacy docs](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/DragandDrop/Concepts/dragsource.html).

```csharp
private class BindableNSDraggingSource : NSDraggingSource
{
	[Export("draggedImage:beganAt:")]
	public void DraggingSessionWillBegin(NSImage session, CGPoint screenPoint)
	{
	}

	[Export("draggedImage:movedTo:")]
	public void DraggingSessionMoved(NSImage session, CGPoint screenPoint)
	{
	}

	[Export("draggedImage:endedAt:operation:")]
	public void DraggingSessionEnded(NSImage session, CGPoint screenPoint, NSDragOperation operation)
	{
	}
}
```

It's a bit ugly but the DraggingSessionEnded method is invoked now when an external drop is completed... yikes.

Reply:
Ah, I did not see these were in the Xamarin class. Wasn't really looking honestly since they weren't in the macOS docs.